File: recipes/agent-memory/python/ci.py
"""Headless integration test for the agent-memory recipe.

Local mode  (default, no secrets): PerSQL(local=":memory:")
  — proves the SDK and MemoryStore API contract. No LLM call.

Remote mode (PERSQL_TOKEN + PERSQL_DATABASE set): claims a fresh branch
  per run for isolation, runs tests, deletes the branch on exit.
  With OPENAI_API_KEY also set: proves the full agent turn end-to-end.

Exit 0 on pass, non-zero on failure. No interactive input.
"""

from __future__ import annotations

import asyncio
import os
import time

from dotenv import load_dotenv

load_dotenv()

PERSQL_TOKEN = os.environ.get("PERSQL_TOKEN")
PERSQL_DATABASE = os.environ.get("PERSQL_DATABASE")
OPENAI_API_KEY = os.environ.get("OPENAI_API_KEY")
RUN_ID = os.environ.get("GITHUB_RUN_ID") or str(int(time.time()))

mode = "remote" if PERSQL_TOKEN else "local"
print(f"[ci] mode={mode}")

from persql import PerSQL
from main import MemoryStore, make_memory_tools  # type: ignore[import]


async def main() -> None:
    cleanup = None

    if PERSQL_TOKEN and PERSQL_DATABASE:
        # Claim a fresh branch per run for isolation.
        parent = PerSQL(token=PERSQL_TOKEN)
        branch_ref = f"ci-{RUN_ID}"
        claimed = parent.database(PERSQL_DATABASE).branches.claim(
            ref=branch_ref, role="admin", ttl_sec=3600
        )
        persql = PerSQL(token=claimed["token"])
        db_path = f"{claimed['namespaceSlug']}/{claimed['databaseSlug']}"
        store = MemoryStore(token=claimed["token"], database=db_path)

        async def _cleanup() -> None:
            parent.database(PERSQL_DATABASE).branches.delete(branch_ref)
            print("[ci] branch cleaned up")

        cleanup = _cleanup
    else:
        client = PerSQL(local=":memory:")
        store = MemoryStore.__new__(MemoryStore)
        store._client = client  # type: ignore[attr-defined]
        store._db = client.database("test/ci")  # type: ignore[attr-defined]
        store._ready = False  # type: ignore[attr-defined]

    try:
        store.init()

        # 1. Save a known memory.
        store.remember(
            name="ci-test-fact",
            description="Widget table schema for CI",
            body="The widgets table has: id INTEGER PRIMARY KEY, name TEXT, price REAL.",
            type="project",
        )
        print("[ci] memory saved")

        # 2. Verify round-trip via recall.
        hits = store.recall("widgets price")
        assert hits, "recall returned no results"
        assert "price REAL" in hits[0]["body"], f"unexpected body: {hits[0]['body']}"
        print("[ci] recall ok")

        # 3. Agent turn — answer from injected memory without a recall tool call.
        if not OPENAI_API_KEY:
            print("[ci] OPENAI_API_KEY not set — skipping agent turn")
        else:
            from openai import AsyncOpenAI
            from agents import (
                Agent,
                OpenAIChatCompletionsModel,
                Runner,
                set_tracing_disabled,
            )

            openai_base_url = os.environ.get("OPENAI_BASE_URL") or None
            openai_model = os.environ.get("OPENAI_MODEL") or "gpt-4o-mini"

            # Custom gateway: speaks /chat/completions, not /responses, and has
            # no platform.openai.com tracing.
            set_tracing_disabled(True)
            client = AsyncOpenAI(api_key=OPENAI_API_KEY, base_url=openai_base_url)
            model = OpenAIChatCompletionsModel(model=openai_model, openai_client=client)

            memories = store.index()
            mem_section = "\n\n".join(
                f"[{m['type']}] {m['name']}\n{m['description']}\n---\n{m['body']}"
                for m in memories
            )

            agent = Agent(
                name="ci-agent",
                model=model,
                instructions=(
                    "You are a helpful assistant.\n\n"
                    f"MEMORIES:\n{mem_section}\n\n"
                    "Answer questions covered above directly without calling any tool first."
                ),
                tools=make_memory_tools(store),
            )

            result = await Runner.run(agent, "What columns does the widgets table have?")
            output = result.final_output or ""
            assert output, "agent returned empty output"
            assert "price" in output.lower(), f"response did not mention price: {output}"
            print(f"[ci] agent turn ok — \"{output[:80]}...\"")

        # 4. Forget and confirm gone.
        store.forget("ci-test-fact")
        after = store.recall("widgets price")
        assert not after, "memory still returned after forget"
        print("[ci] forget ok")

        print(f"[ci] PASS ({mode})")

    finally:
        if cleanup:
            await cleanup()


if __name__ == "__main__":
    asyncio.run(main())

File: recipes/agent-memory/typescript/ci.ts
import "dotenv/config";
import assert from "node:assert/strict";
import OpenAI from "openai";
import {
  Agent,
  run,
  setDefaultOpenAIClient,
  setOpenAIAPI,
  setTracingDisabled,
} from "@openai/agents";
import { PerSQL } from "@persql/sdk";
import { MemoryStore, makeMemoryTools } from "./memory.js";

const persqlToken = process.env.PERSQL_TOKEN;
const persqlDatabase = process.env.PERSQL_DATABASE;
const openaiKey = process.env.OPENAI_API_KEY;
// Optional OpenAI-compatible gateway. Unset → the SDK's default OpenAI API.
const openaiBaseUrl = process.env.OPENAI_BASE_URL;
const openaiModel = process.env.OPENAI_MODEL || "gpt-4o-mini";
const runId = process.env.GITHUB_RUN_ID ?? `local-${Date.now()}`;

const mode = persqlToken ? "remote" : "local";
console.log(`[ci] mode=${mode}`);

// Remote: claim a fresh branch per run for isolation, delete it on exit.
// Local:  in-process :memory: SQLite, no network, no cost.
let persql: PerSQL;
let store: MemoryStore;
let cleanup: (() => Promise<void>) | undefined;

if (persqlToken && persqlDatabase) {
  const parent = new PerSQL({ token: persqlToken });
  const branchRef = process.env.PERSQL_BRANCH_REF ?? `ci-${runId}`;
  const claimed = await parent.database(persqlDatabase).branches.claim({
    ref: branchRef,
    role: "admin",
    ttlSec: 3600,
  });
  persql = new PerSQL({ token: claimed.token });
  store = new MemoryStore(
    persql.database(`${claimed.namespaceSlug}/${claimed.databaseSlug}`)
  );
  cleanup = async () => {
    await parent.database(persqlDatabase).branches.delete(branchRef);
    console.log("[ci] branch cleaned up");
  };
} else {
  persql = new PerSQL({ local: ":memory:" });
  store = new MemoryStore(persql.database("test/ci"));
}

try {
  await store.init();

  // 1. Save a known memory.
  await store.remember({
    name: "ci-test-fact",
    description: "Widget table schema for CI",
    type: "project",
    body: "The widgets table has: id INTEGER PRIMARY KEY, name TEXT, price REAL.",
  });
  console.log("[ci] memory saved");

  // 2. Round-trip via recall.
  const hits = await store.recall("widgets price");
  assert.ok(hits.length > 0, "recall returned no results");
  assert.ok(hits[0].body.includes("price REAL"), `unexpected body: ${hits[0].body}`);
  console.log("[ci] recall ok");

  // 3. Agent turn — answer from injected memory, no recall tool call.
  // @openai/agents reads OPENAI_API_KEY from the environment automatically.
  if (!openaiKey) {
    console.log("[ci] OPENAI_API_KEY not set — skipping agent turn");
  } else {
    if (openaiBaseUrl) {
      // Custom gateway: cast needed (@openai/agents bundles a nested openai
      // sub-dep). Gateways speak /chat/completions, not /responses, and have
      // no platform.openai.com tracing.
      setDefaultOpenAIClient(
        new OpenAI({ apiKey: openaiKey, baseURL: openaiBaseUrl }) as unknown as Parameters<
          typeof setDefaultOpenAIClient
        >[0]
      );
      setOpenAIAPI("chat_completions");
      setTracingDisabled(true);
    }

    const memories = await store.index();
    const memSection = memories
      .map((m) => `[${m.type}] ${m.name}\n${m.description}\n---\n${m.body}`)
      .join("\n\n");

    const agent = new Agent({
      name: "ci-agent",
      model: openaiModel,
      instructions: `You are a helpful assistant.\n\nMEMORIES:\n${memSection}\n\nAnswer questions covered above directly without calling any tool first.`,
      tools: makeMemoryTools(store),
    });

    const result = await run(agent, "What columns does the widgets table have?");
    assert.ok(result.finalOutput, "agent returned empty output");
    assert.ok(
      result.finalOutput.toLowerCase().includes("price"),
      `response did not mention price: ${result.finalOutput}`
    );
    console.log(`[ci] agent turn ok — "${result.finalOutput.slice(0, 80)}..."`);

    // 3b. Tool selection (write) — state a NEW fact in plain language and
    // require the model to persist it by *calling* remember_memory. The
    // store only mutates through that tool, so a row appearing proves the
    // model chose the tool (not just narrated saving). The token sku_7f3a
    // is unguessable, so the later recall can't pass by the model guessing.
    const writer = new Agent({
      name: "ci-writer",
      model: openaiModel,
      instructions:
        "You take notes. When the user shares a fact worth keeping, save it " +
        "with the remember_memory tool using a short kebab-case name.",
      tools: makeMemoryTools(store),
    });
    await run(
      writer,
      "Please remember this: the orders table has columns id, total, status, and sku_7f3a."
    );
    const persisted = (await store.index()).find((m) =>
      m.body.toLowerCase().includes("sku_7f3a")
    );
    assert.ok(persisted, "agent did not persist the fact via remember_memory");
    console.log(`[ci] tool-call remember ok — saved "${persisted.name}"`);

    // 3c. Recall. First prove the mechanism deterministically: the fact the
    // writer just saved must be retrievable by search. This is the real
    // regression guard — it fails loudly if recall breaks.
    const searchHits = await store.recall("orders table columns");
    assert.ok(
      searchHits.some((h) => h.body.toLowerCase().includes("sku_7f3a")),
      "recall() did not return the just-saved orders fact — search is broken"
    );

    // Then the agentic turn — a fresh agent with no memory injected, expected
    // to call recall_memory and surface the fact. We assert on the tool's
    // output, not the model's prose: a small model that paraphrases (or, on an
    // off run, skips the tool) shouldn't flake the build when recall itself
    // demonstrably works above. But if the model *does* call recall_memory,
    // its result must contain the fact — a tool wired to return nothing is a
    // real bug, not flakiness.
    const reader = new Agent({
      name: "ci-reader",
      model: openaiModel,
      instructions:
        "You have no prior knowledge of the user's data. Always call " +
        "recall_memory to look up what the user asks about before answering.",
      tools: makeMemoryTools(store),
    });
    const recalled = await run(reader, "What columns does the orders table have?");

    const toolNames: string[] = [];
    const toolOutputs: string[] = [];
    for (const item of recalled.newItems) {
      if (item.type === "tool_call_item") {
        const name = (item.rawItem as { name?: string }).name;
        if (name) toolNames.push(name);
      } else if (item.type === "tool_call_output_item") {
        if (item.output != null) toolOutputs.push(String(item.output));
      }
    }

    if (toolNames.includes("recall_memory")) {
      assert.ok(
        toolOutputs.some((o) => o.toLowerCase().includes("sku_7f3a")),
        "recall_memory was called but returned no matching fact — recall is broken"
      );
    }

    const recalledOut = (recalled.finalOutput ?? "").toLowerCase();
    if (
      recalledOut.includes("sku_7f3a") ||
      toolOutputs.some((o) => o.toLowerCase().includes("sku_7f3a"))
    ) {
      console.log("[ci] tool-call recall ok");
    } else {
      console.log(
        "[ci] note: reader skipped recall_memory this run (model " +
          "nondeterminism); recall verified directly above"
      );
    }
  }

  // 4. Forget and confirm gone.
  await store.forget("ci-test-fact");
  const after = await store.recall("widgets price");
  assert.ok(after.length === 0, "memory still returned after forget");
  console.log("[ci] forget ok");

  console.log(`[ci] PASS (${mode})`);
} finally {
  await cleanup?.();
}
